src/Type/Php/CurlGetinfoFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,198 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\ConstantType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\FloatType;
+use PHPStan\Type\IntegerType;
+use PHPStan\Type\NullType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function count;
+use function strtolower;
+
+final class CurlGetinfoFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function __construct(
+		private ReflectionProvider $reflectionProvider,
+	)
+	{
+	}
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return strtolower($functionReflection->getName()) === 'curl_getinfo';
+	}
+
+	/**
+	 * @throws ShouldNotHappenException
+	 */
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (count($functionCall->getArgs()) < 1) {
+			return ParametersAcceptorSelector::selectSingle(
+				$functionReflection->getVariants(),
+			)->getReturnType();
+		}
+
+		if (count($functionCall->getArgs()) > 1) {
+			$componentType = $scope->getType($functionCall->getArgs()[1]->value);
+
+			if (!$componentType instanceof ConstantType) {
+				return $this->createAllComponentsReturnType();
+			}
+
+			$componentType = $componentType->toInteger();
+			if (!$componentType instanceof ConstantIntegerType) {
+				throw new ShouldNotHappenException();
+			}
+		} else {
+			$componentType = new ConstantIntegerType(-1);
+		}
+
+		if ($componentType->getValue() === -1) {
+			return $this->createAllComponentsReturnType();
+		}
+
+		$componentTypesPairedConstants = [
+			'CURLINFO_EFFECTIVE_URL' => new StringType(),
+			'CURLINFO_FILETIME' => new IntegerType(),
+			'CURLINFO_TOTAL_TIME' => new FloatType(),
+			'CURLINFO_NAMELOOKUP_TIME' => new FloatType(),
+			'CURLINFO_CONNECT_TIME' => new FloatType(),
+			'CURLINFO_PRETRANSFER_TIME' => new FloatType(),
+			'CURLINFO_STARTTRANSFER_TIME' => new FloatType(),
+			'CURLINFO_REDIRECT_COUNT' => new IntegerType(),
```

Comment:
Please be more memory usage friendly, assign variables with these types and then use the variables instead of creating new instance every time.